src/data/pfun.lean

```diff
@@ -448,4 +448,45 @@ ext $ λ _ _, by simp only [comp_apply, part.bind_comp]
 lemma coe_comp (g : β → γ) (f : α → β) : ((g ∘ f : α → γ) : α →. γ) = (g : β →. γ).comp f :=
 ext $ λ _ _, by simp only [coe_val, comp_apply, part.bind_some]
 
+/-- Product of partial functions. -/
+def prod_lift (f : α →. β) (g : α →. γ) : α →. β × γ :=
+λ x, ⟨(f x).dom ∧ (g x).dom, λ h, ((f x).get h.1, (g x).get h.2)⟩
+
+@[simp] lemma dom_prod_lift (f : α →. β) (g : α →. γ) :
+  (f.prod_lift g).dom = {x | (f x).dom ∧ (g x).dom} := rfl
+
+lemma get_prod_lift (f : α →. β) (g : α →. γ) (x : α) (h) :
+  (f.prod_lift g x).get h = ((f x).get h.1, (g x).get h.2) := rfl
+
+@[simp] lemma prod_lift_apply (f : α →. β) (g : α →. γ) (x : α) :
+  f.prod_lift g x = ⟨(f x).dom ∧ (g x).dom, λ h, ((f x).get h.1, (g x).get h.2)⟩ := rfl
+
+lemma mem_prod_lift {f : α →. β} {g : α →. γ} {x : α} {y : β × γ} :
+  y ∈ f.prod_lift g x ↔ y.1 ∈ f x ∧ y.2 ∈ g x :=
+by { simp [prod_lift], tidy }
```

Comment:
I guess this `tidy` is pretty quick, but it would be nice to squeeze it anyway: a priori this has nonterminal simps in it

Reply:
I don't understand the idea that "`tidy` has nonterminal simps in it". Isn't it considered a terminating tactic? Like sure `simp` has simps in it too, but that doesn't make it illegal.

Reply:
Sure, but remember that the tidy might end up doing things like `simp, split, dsimp, simp`; so if new simp lemmas are added, an application of `tidy` can now break - which is why things like this can be hard to maintain

Reply:
Sure, but what `tidy` does internally is irrelevant. Externally, you have one terminal `tidy` call.

Reply:
In what sense is it irrelevant? If I define a custom tactic which does `simp, intros, split, refl`, then using it terminally would still amount to using nonterminal simps - it's precisely what it does internally that is the issue

Reply:
Let's continue on [Zulip](https://leanprover.zulipchat.com/#narrow/stream/113488-general).